### ISSUE TITLE
New Rule Proposal: class-members-in-constructor

### ISSUE THREAD

**Issue by stevendesu** [eslint/eslint#11540]
**Please describe what the rule should do:**

When working with ES6 classes, members are created on-the-fly when referenced (since classes extend the Object prototype). This allows for simple typos to go unnoticed. This rule should require that any class members (variables referenced via `this.xxx`) are originally declared in the constructor, preventing attempts to access invalid members.

**What category of rule is this? (place an "X" next to just one item)**

I would consider this rule belonging to [Best Practices](https://eslint.org/docs/rules/#best-practices)

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

Simple typo referencing undefined variable:
```js
class Greeter
{
    constructor(name)
    {
        this.myName = name;
    }
    sayHello()
    {
        console.log("Hello, " + this.myname);
    }
}
var bob = new Greeter("bob");
bob.sayHello(); // Prints "Hello, undefined"
```

Race condition where initialization of variable can't be known:
```js
class Race
{
    loadRemoteFile()
    {
        fetch("http://www.google.com").then(resp =>
        {
            resp.text().then(text => this.remoteFile = text);
        });
    }
    getFileSize()
    {
        return this.remoteFile.length;
    }
}
var google = new Race();
google.loadRemoteFile();
setTimeout(function()
{
    console.log(google.getFileSize()); // May print a Number, may throw error "cannot get property length of undefined"
}, 50);
```

Race condition resolved by defining member in constructor:
```js
class Race
{
    constructor()
    {
        this.remoteFile = "";
    }
    loadRemoteFile()
    {
        fetch("http://www.google.com").then(resp =>
        {
            resp.text().then(text => this.remoteFile = text);
        });
    }
    getFileSize()
    {
        return this.remoteFile.length;
    }
}
var google = new Race();
google.loadRemoteFile();
setTimeout(function()
{
    console.log(google.getFileSize()); // Always prints a Number (sometimes 0)
}, 50);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is effectively an ES6 class corollary for [no-use-before-define](https://eslint.org/docs/rules/no-use-before-define), extending it to ensure that all class members are defined before any class methods are called (must be defined in the constructor)

There is a potential edge case for class methods **called by** the constructor, such as:

```js
class EdgeCase
{
    constructor()
    {
        this.useVariable();
        this.variable = 3;
    }
    useVariable()
    {
        console.log(this.variable);
    }
}
```

Detecting this should be possible, but may likely be much more complex. Any methods called within the constructor should not be allowed to mention variables declared on a lower line in the constructor.

**Are you willing to submit a pull request to implement this rule?**

I would be willing, but I can't guarantee I would be able to do so quickly. There's a lot of learning I'd have to do in order to understand how ESLint works.

** Minor Edit **

Upon realizing that `resp.text()` returns a Promise, I updated the race condition example

**Comment by stevendesu** [comment]
Putting some thought into the edge case, this could actually be trivially solved by not allowing function calls before variable declarations. The same way it's best practice to have all imports at the top of a file or all variable declarations at the top of a file, why not have all class member declarations at the top of the constructor?

```js
class EdgeCase
{
    constructor()
    {
        this.useVariable(); // ESLINT ERROR: class-members-declared-first or something like that
        this.variable = 3;
}
```
